src/devices/sound/nes_apu.cpp

```diff
@@ -724,17 +735,14 @@ void nesapu_device::write(offs_t address, u8 value)
 
 void nesapu_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
 {
-	int accum;
+	stream_buffer::sample_t accum = 0.0;
 	auto &output = outputs[0];
 
 	for (int sampindex = 0; sampindex < output.samples(); sampindex++)
 	{
-		accum = apu_square(&m_APU.squ[0]);
-		accum += apu_square(&m_APU.squ[1]);
-		accum += apu_triangle(&m_APU.tri);
-		accum += apu_noise(&m_APU.noi);
-		accum += apu_dpcm(&m_APU.dpcm);
+		accum = m_square_lut[apu_square(&m_APU.squ[0])][apu_square(&m_APU.squ[1])];
+		accum += m_tnd_lut[apu_triangle(&m_APU.tri)][apu_noise(&m_APU.noi)][apu_dpcm(&m_APU.dpcm)];
```

Comment:
Do we need to do these by table lookup? Why not just use the exact formula?

Let's presume we want the table for performance reasons. According to the docs you're using the formulas above to calculate via the less exact approximation. For that we only need a 31-entry table and a 203-entry table with indices [squ0 + squ1] and [3 * triangle + 2 * noise + dpmc] here.

On the other hand, if you indeed want to calculate the full table of exact values (32k entries in the m_tnd_lut case) then the formulas have the wrong numerators and the tnd channels need divisors.